### DISCUSSION TITLE
How to Fix Input Matrices in Yalmip During Optimization?

### DISCUSSION THREAD

**Discussion by maniente** [yalmip/YALMIP#1424] [Optimization modelling]
Hello everyone,

I'm working on a Model Predictive Control (MPC) problem using Yalmip, and I've run into an issue regarding input matrices. Specifically, I have the matrices Am, Bm, and E that I want to keep fixed during the optimization process. These matrices are recalculated at each instant of the simulation based on a time-varying vector v. The problem is that I noticed the matrix Am is being optimized and changes during the simulation, whereas Bm and E remain fixed as intended.

Is there a way to ensure that the matrix Am remains fixed during the optimization, or am I missing something in my implementation?

Here is the relevant part of my code:
Controller
```
u = sdpvar(repmat(nu, 1, N), repmat(1, 1, N));
x = sdpvar(repmat(nx, 1, N+1), repmat(1, 1, N+1));
r = sdpvar(repmat(ny, 1, N+1), repmat(1, 1, N+1));
Am = sdpvar(nx, nx);
Bm = sdpvar(nx, nu);
E = sdpvar(nx, 1);

C = [1 0 0 0; 0 0 1 0];

constraints = [];
objective = 0;

for k = 1:N
    objective = objective + (C*x{k} - r{k})' * Q * (C*x{k} - r{k}) + u{k}' * R * u{k};
    constraints = [constraints, x{k+1} == Am * x{k} + Bm * u{k} + E];
    constraints = [constraints, [-10; -10] <= [1 0; 1 1] * u{k} <= [10; 10]];
}

objective = objective + (C*x{N+1} - r{N+1})' * Q * (C*x{N+1} - r{N+1});
parameters_in = {x{1}, [r{:}], Am, Bm, E};
solutions_out = {[u{:}], [x{:}]};

options = sdpsettings('solver', 'mosek');
controller = optimizer(constraints, objective, options, parameters_in, solutions_out);

```
Example call to controller
```
future_r = Rs(:, j:j+N);
[Av, Bv, Ev] = matrici(v(j), ind, Ts);
inputs = {xx(:, j), future_r, Av, Bv, Ev};
[solutions, ~] = controller{inputs};`

```

Any advice or suggestions would be greatly appreciated.

Thank you!

**Comment by johanlofberg** [comment]
You are telling YALMIP the parameter A is symmetric (which I don't think it is) so it will get completely messed up when sending non-symmetric data

**Reply by maniente** [reply]
Oops, I hadn't noticed that at all. Thank you!
